pandas/tests/computation/test_eval.py

```diff
@@ -441,87 +441,123 @@ def check_compound_invert_op(self, lhs, cmp1, rhs):
     def ex(self, op, var_name="lhs"):
         return f"{op}{var_name}"
 
-    def test_frame_invert(self):
+    @pytest.mark.parametrize(
+        "lhs",
+        [
+            DataFrame(randn(5, 2)),
+            DataFrame(randint(5, size=(5, 2))),
+            DataFrame(rand(5, 2) > 0.5),
+            DataFrame({"b": ["a", 1, 2.0], "c": rand(3) > 0.5}),
+        ],
+    )
+    def test_frame_invert(self, lhs):
         expr = self.ex("~")
+        dtype = lhs.T.dtypes[0]
 
-        # ~ ##
-        # frame
-        # float always raises
-        lhs = DataFrame(randn(5, 2))
-        if self.engine == "numexpr":
-            with pytest.raises(NotImplementedError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
-        else:
-            with pytest.raises(TypeError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
+        is_engine_numexpr = self.engine == "numexpr"
 
-        # int raises on numexpr
-        lhs = DataFrame(randint(5, size=(5, 2)))
-        if self.engine == "numexpr":
-            with pytest.raises(NotImplementedError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
-        else:
-            expect = ~lhs
-            result = pd.eval(expr, engine=self.engine, parser=self.parser)
-            tm.assert_frame_equal(expect, result)
+        if dtype in ["float", "object"]:
+            pytest.skip(f"dtype of '{dtype}' will always raise.")
+        elif dtype == "int" and is_engine_numexpr:
+            pytest.skip("'numexpr' will raise with dtype of 'int'")
 
-        # bool always works
-        lhs = DataFrame(rand(5, 2) > 0.5)
         expect = ~lhs
         result = pd.eval(expr, engine=self.engine, parser=self.parser)
         tm.assert_frame_equal(expect, result)
 
-        # object raises
-        lhs = DataFrame({"b": ["a", 1, 2.0], "c": rand(3) > 0.5})
-        if self.engine == "numexpr":
-            with pytest.raises(ValueError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
-        else:
-            with pytest.raises(TypeError):
-                result = pd.eval(expr, engine=self.engine, parser=self.parser)
+    @pytest.mark.parametrize(
```

Comment:
can you just use the engine fixuture, these if/thens are *very* confusing

Reply:
I think you might be better off adding *more* parameterization here on the input which cleanly enumerates the cases; then on a speciifc input you can for example put the error. 

essentially we don't want *any* if/thens in the body of the function.

I know this is tricky, this module is pretty hard and the existing patterns don't work very well.

Reply:
@jreback 

> can you just use the engine fixuture, these if/thens are _very_ confusing

I don't see how the ```engine``` fixture will prevent the use of ```if/else``` statements, because from what I understand it will only validate if ```_USE_NUMEXPR``` is set to ```True```, there is no ```dtype``` validation anywhere. or I am missing something? (probably missing something)

> I think you might be better off adding _more_ parameterization here on the input which cleanly enumerates the cases; then on a speciifc input you can for example put the error.
> 

What have you thought about, I tried putting ```(lhs, engine_name, error_type)``` (3 parameters), but the code looked sloppy and similar to [my_first_calculator.py](https://github.com/AceLewis/my_first_calculator.py) 

> essentially we don't want _any_ if/thens in the body of the function.
> 

At all? going to be extremely tricky, because the error that get raised is depends both on the ```dtype``` of the ```DataFrame```/```Series```, the only thing (right now) that I can think about is creating an external function to do that.

> I know this is tricky, this module is pretty hard and the existing patterns don't work very well.